chain/network/src/tcp.rs

```diff
@@ -118,15 +118,74 @@ impl Stream {
     }
 }
 
-pub(crate) struct Listener(tokio::net::TcpListener);
+/// In production code ListenerAddr is isomorphic to std::net::SocketAddr.
+/// In tests it additionally allows to "reserve" a random unused TCP port:
+/// * it allows to avoid race conditions in tests which require a dedicated TCP port to spawn a
+///   node on (and potentially restart it every now and then).
+/// * it is implemented by usign SO_REUSEPORT socket option (do not confuse with SO_REUSEADDR),
+///   which allows multiple sockets to share a port. new_for_test() creates a socket and binds
+///   it to a random unused local port (without starting a TCP listener).
+///   This socket won't be used for anything but telling the OS that the given TCP port is in use.
+///   However thanks to SO_REUSEPORT we can create another socket bind it to the same port
+///   and make it a listener.
+/// * The drawback of this solution that it is hard to detect a situation in which multiple
+///   listener sockets in test are bound to the same port. TODO(gprusak): we can prevent creating
+///   multiple listeners for a single port within the same process by implementing a global register
+///   of ports in use.
+#[derive(Clone)]
+pub struct ListenerAddr {
+    /// A test-only guard ensuring that OS considers the given TCP port to be in use (in prod,
+    /// guard should be always None).
+    guard: Arc<Option<tokio::net::TcpSocket>>,
+    addr: std::net::SocketAddr,
+}
 
-impl Listener {
-    // TODO(gprusak): this shouldn't be async. It is only
-    // because TcpListener accepts anything that asynchronously resolves to SocketAddr.
-    pub async fn bind(addr: std::net::SocketAddr) -> std::io::Result<Self> {
-        Ok(Self(tokio::net::TcpListener::bind(addr).await?))
+impl fmt::Debug for ListenerAddr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.addr.fmt(f)
+    }
+}
+
+impl std::ops::Deref for ListenerAddr {
+    type Target = std::net::SocketAddr;
+    fn deref(&self) -> &Self::Target {
+        &self.addr
+    }
+}
+
+impl ListenerAddr {
+    pub fn new(addr: std::net::SocketAddr) -> Self {
+        assert!(addr.port()!=0, "using an arbitrary port for the tcp::Listener is allowed only in tests and only via new_for_test() method");
```

Comment:
Did `rustfmt` leave it like this?

```suggestion
        assert!(
            addr.port() != 0, 
            "using an arbitrary port for the tcp::Listener is allowed only \
             in tests and only via new_for_test() method"
         );
```

Reply:
cargo fmt did